### DISCUSSION TITLE
How to program the CM4?

### DISCUSSION THREAD

**Discussion by ogg1e** [mebs/CM4-NAS#27] [Q&A] [answered]
- What are the steps to program the eMMC on this board?
- Do you need to set a jumper like some other IO boards?
- Do you need to apply power to the 5v connector when programming, or is usb power enough?

**Comment by ogg1e** [comment]
@mebs ? Can you answer my questions possibly?

**Comment by mebs** [comment, marked as answer]
Hi, yes you need two jumpers. One to set the USB as slave and one to put the CM4 in programming mode.
To make the CM4 act as a slave USB device, you need to short these two pins, the silkscreen should help:
![image](https://user-images.githubusercontent.com/2614134/152791947-187485ec-08c4-412d-a432-68468ad7c3eb.png)

To transfer boot from the eMMC to rpi boot via USB (put the CM4 in programming mode) you need to bring the `nRPI_BOOT` pin to ground, which can be done by shorting these pins on the bottom of the board:
![image](https://user-images.githubusercontent.com/2614134/152792363-c755b960-3659-45c0-a25c-a137656015e8.png)

**Reply by ogg1e** [reply]
Thanks @mebs! 

About powering the CM4 during flashing.  Do I need to apply 5v to the power connector? Or maybe 5v to the GPIO pins? Or can it be powered over usb?

**Reply by mebs** [reply]
Ah yes, skipped that part of the question.
The USB VBUS from the master (the PC) is not routed to the CM4's VCC. You have to provide power through the power header on the board, as if you were running it.

**Reply by mebs** [reply]
(Once the two jumpers are in place and power is provided, the steps to program the eMMC are the same that are described in all the tutorials on the internet, of course.)
